org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -0,0 +1,1296 @@
+    }
+
+    private void initServer() {
+        logger.debug("JSSEngine: initServer()");
+
+        // The only time cert and key are required are when we're creating a
+        // server SSLEngine.
+        if (cert == null || key == null) {
+            throw new IllegalArgumentException("JSSEngine: must be initialized with server certificate and key!");
+        }
+
+        logger.debug("JSSEngine.initServer(): " + cert);
+        logger.debug("JSSEngine.initServer(): " + key);
+
+        // Configure SSL server with the given certificate and its private
+        // key.
+        if (SSL.ConfigServerCert(ssl_fd, cert, key) == SSL.SECFailure) {
+            throw new RuntimeException("Unable to initialize server with cert and key: " + errorText(PR.GetError()));
+        }
+        session.setLocalCertificates(new PK11Cert[]{ cert } );
+
+        // Create a small session cache.
+        //
+        // TODO: Make this configurable.
+        if (SSL.ConfigServerSessionIDCache(1, 100, 100, null) == SSL.SECFailure) {
+            throw new RuntimeException("Unable to configure server session cache " + errorText(PR.GetError()));
+        }
+
+        // Update handshake status; client initiates connection, so wait
+        // for unwrap on the server end.
+        handshake_state = SSLEngineResult.HandshakeStatus.NEED_UNWRAP;
+
+        // Only specify these on the server side as they affect what we
+        // want from the remote peer in NSS. In the server case, this is
+        // client auth, but if we were to set these on the client, it would
+        // affect server auth.
+        if (SSL.OptionSet(ssl_fd, SSL.REQUEST_CERTIFICATE, want_client_auth || need_client_auth ? 1 : 0) == SSL.SECFailure) {
+            throw new RuntimeException("Unable to configure SSL_REQUEST_CERTIFICATE option: " + errorText(PR.GetError()));
+        }
+
+        if (SSL.OptionSet(ssl_fd, SSL.REQUIRE_CERTIFICATE, need_client_auth ? 1 : 0) == SSL.SECFailure) {
+            throw new RuntimeException("Unable to configure SSL_REQUIRE_CERTIFICATE option: " + errorText(PR.GetError()));
+        }
+
+        step_handshake = true;
+    }
+
+    private void applyCiphers() {
+        logger.debug("JSSEngine: applyCiphers()");
+        // Enabled the ciphersuites specified by setEnabledCipherSuites(...).
+        // When this isn't called, enabled_ciphers will be null, so we'll just
+        // use whatever is enabled by default.
+        if (enabled_ciphers == null) {
+            return;
+        }
+
+        // We need to disable the suite if it isn't present in the list of
+        // suites above. Be lazy about it for the time being and disable all
+        // cipher suites first.
+        for (SSLCipher suite : SSLCipher.values()) {
+            if (SSL.CipherPrefSet(ssl_fd, suite.getID(), false) == SSL.SECFailure) {
+                logger.debug("Unable to set cipher suite preference for " + suite.name() + ": " + errorText(PR.GetError()));
```

Comment:
Do we ignore this failure? I think the logging level should be WARN.

Reply:
Yeah, we ignore this failure. The client can pass all sorts of weird cipher suites to us, which NSS won't necessarily understand. Like the [KRB5 cipher suites](https://tools.ietf.org/html/rfc2712). :D